e2e/investigation.md

```diff
@@ -0,0 +1,103 @@
+
+Because of the nature of the e2e tests there are inherent issues:
+
+    Not every test can run parallel to other tests in the same metal3-dev-env instance e.g.
+    pivoting needs to be executed before re-pivoting. Other example Updating the management cluster or
+    other objects in the cluster during an "upgrade test" might cause stability issues for other
+    tests in the same environment.
+
+    At the moment the cleanup process for e2e tests in CAPM3 is not test specific.
+    There is no cleanup process implemented for individual e2e tests that would restore the dev-env to a state
+    from where other e2e tests could be started, this is partly the case because the restoration process is highly
+    dependent on the e2e test that was executed as some tests affect the configuration of the whole dev-env.
+
+It has become clear that there are multiple ways to achieve the parallel test execution
+
+## Possible solutions
+
+**Option 1 (pool of dev-envs)**:
+
+    This solution requires the bootstrapping of separate dev-env instances
+    (both management cluster and target cluster) for individual e2e test cases.
+    (CON: it scales poorly and it would cause problems with locally running tests on a developer's machine)
+
+**Option 2 (solve it in CI)**:
+
+    This approach would require minimal modification to the CAPM3 Makefile and
+    additional some additional logic to the test code to implement the ability to selectively execute e2e test
+    specs (similarly how it is in dev-env integration tests). Option 3 also includes the individually executable tests.
+
+    This approach would basically be implemented by running multiple e2e test jobs in CI that would only test individual e2e
+    test cases. (CON: this would still require multiple hosts in CityCloud and it would be still slow on local dev-envs).
+
+**Option3 (GO approach)**:
+
+Ginkgo test specs can be executed in a parallel fashion as described [here](https://onsi.github.io/ginkgo/#parallel-specs).
+
+    Each individual test case should have it's own cleanup process that reverts the modifications
+    done by the test execution.
+
+    The cleanup process should aim to avoid re-installation of the
+    dev-env. In some cases like pivoting the cleanup process could be implemented as an other
+    test spec/case e.g "re-pivoting".
+
+    Test specs that are executed in parallel mode should be grouped separately from the ones executed sequentially as
+    not all e2e test cases work well together. The tests that are grouped together as parallelly executable would be
+    executed by Ginkgo's native parallel execution feature. The parallelly executed tests would run at the same time
+    in different namespaces on the same developer environment (same management cluster and same managed cluster(s)).
+
+    All the test cases should be able run individually e.g "make test-e2e pivoting update" etc...
```

Comment:
It is mentioned above that tests should be grouped since they may need to run sequentially. (Pivoting - Repivoting). So how we can make all the test being able to run individually?

Reply:
This question connected to the previous one so I will answer both here. 
There are multiple things that have to be done.

Both of the questions relate to Option3 that is might be not enough by itself as I described. It might require either option 1 or 2 to be implemented alongside with it in the future.

Answer to your question:

1. We would need to implement an environment variable based selector logic to enable-disable groups of tests, I think these "groups" of tests could even be in different "Describe" blocks. 

2. We need at least 2 separate GO test execution command in the makefile, they could stay under the same "make test-e2e" command. One  would be the sequential execution command and the others would commands that execute the group of tests that could run parallel to each other.

Reply:
I'm happy with your answer of the question 2. However, there are something that still make me confused regarding the question 1. My concern is that we will hardly find a pair of tests that will not interfere each other if we run them at the same time on the same nodes. Could you explain a bit more about how to avoid this conflict by using different Describe blocks? 

> Both of the questions relate to Option3 that is might be not enough by itself as I described. It might require either option 1 or 2 to be implemented alongside with it in the future.

What do you mean by `not enough`? Does the option 3 not provide the parallel execution as we expect?

Reply:
Answer for the first question:

- It is true that at the moment we have just a few tests and they are mostly disruptive but it might change in the future and I think the goal is to make a maintainable e2e test process so we have to prepare for the future.
- The use of separate describe blocks is just an additional step, I think it would nicely group tests together so this step by itself would not solve the parallelization. 

Answer for second question:

Option 3 would provide a structure for e2e tests that is maintainable and more controllable from execution point of view. I think option 3 has to be implemented in any case but it would not be enough by itself and I think the answer will be that we have to implement option 2 additionally.

Reply:
Ok, I see. Thanks for your answer, @Rozzii. Now I have better understanding about this study. More comments will be added. Please check